<a href="https://colab.research.google.com/github/AycaCicek/Turkcell-GelecegiYazanlar_Notlar/blob/main/%C3%87oklu_Do%C4%9Frusal_Regresyon.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd 
data = pd.read_csv("/content/Advertising.csv")
data.head()

Unnamed: 0     TV  radio  newspaper  sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

In [2]:
df = data.iloc[:,1:len(data)]
df.head()

TV  radio  newspaper  sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

In [3]:
X=df.drop("sales",axis=1)
y=df["sales"]

In [4]:
X.head()

TV  radio  newspaper
0  230.1   37.8       69.2
1   44.5   39.3       45.1
2   17.2   45.9       69.3
3  151.5   41.3       58.5
4  180.8   10.8       58.4

In [5]:
y.head()

0    22.1
1    10.4
2     9.3
3    18.5
4    12.9
Name: sales, dtype: float64

A) STATSMODELS İLE MODEL KURMA 

In [8]:
import statsmodels.api as sm

In [10]:
lm =sm.OLS(y,X)
model=lm.fit()

In [11]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                  sales   R-squared (uncentered):                   0.982
Model:                            OLS   Adj. R-squared (uncentered):              0.982
Method:                 Least Squares   F-statistic:                              3566.
Date:                Sat, 22 Oct 2022   Prob (F-statistic):                   2.43e-171
Time:                        08:51:26   Log-Likelihood:                         -423.54
No. Observations:                 200   AIC:                                      853.1
Df Residuals:                     197   BIC:                                      863.0
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
TV             0.0538      0.001     40.507      0.000       0.051       0.056
radio          0.2222      0.009     23.595      0.000       0.204       0.241
newspaper      0.0168      0.007      2.517      0.013       0.004       0.030
==============================================================================
Omnibus:                        5.982   Durbin-Watson:                   2.038
Prob(Omnibus):                  0.050   Jarque-Bera (JB):                7.039
Skew:                          -0.232   Prob(JB):                       0.0296
Kurtosis:                       3.794   Cond. No.                         12.6
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

* Bağımsız değişkenlerin bağımlı değişkeni açıklama yüzdesi R*2 değeri yüksektir.
* Şişmenin önüne geçebilmek için düzeltilmiş R*2'ye bakmak lazım.
* Model anlamlıdıdr...(F değeri)
* P>|t| değerleri 0,05'ten küçük oldukları için modelde kullanılan değişkenlerin analmalı oludğunu söyleriz.

--> KATSAYI YORUMLAMA ÖRNEK:
1.   TV: Diğer değişkenler sabit tutulduğunda TV harcamalarında meydana gelecek 1 brmlik artış sales değişkeninde ortalama 0.0538 artışa sebep olacaktır.

B) SCIKIT LEARN İLE MODEL KURMA

In [15]:
from sklearn.linear_model import LinearRegression

In [16]:
lm =LinearRegression()

In [20]:
model = lm.fit(X,y)

In [21]:
model.intercept_

2.938889369459412

In [22]:
model.coef_

array([ 0.04576465,  0.18853002, -0.00103749])

TAHMİN 
* FORMÜL
> sales = 2.938889369459412 + 0.04576465 * TV + 0.18853002 * radio + (-0.00103749 * newspaper)



In [26]:
sorgu_tv = float(input("TV harcaması:"))
sorgu_radyo = float(input("Radyo harcaması:"))
sorgu_gazete = float(input("Gazete harcaması:"))

sales = "2.938889369459412 + 0.04576465 * {} + 0.18853002 * {} - 0.00103749 * {}".format(sorgu_tv,sorgu_radyo,sorgu_gazete)

tahmin = 2.938889369459412 + 0.04576465 * sorgu_tv + 0.18853002 * sorgu_radyo + (-0.00103749 * sorgu_gazete)

print(f"""
formül: {sales} 
satış tahmini: {tahmin}""")

TV harcaması:10
Radyo harcaması:15
Gazete harcaması:20

formül: 2.938889369459412 + 0.04576465 * 10.0 + 0.18853002 * 15.0 - 0.00103749 * 20.0 
satış tahmini: 6.2037363694594125


In [27]:
yeni_veri=[[10],[15],[20]]
yeni_veri=pd.DataFrame(yeni_veri).T
yeni_veri

0   1   2
0  10  15  20

In [28]:
model.predict(yeni_veri)

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([6.20373622])

In [29]:
from sklearn.metrics import mean_squared_error

In [30]:
y.head()

0    22.1
1    10.4
2     9.3
3    18.5
4    12.9
Name: sales, dtype: float64

In [31]:
model.predict(X)[0:10]

array([20.52397441, 12.33785482, 12.30767078, 17.59782951, 13.18867186,
       12.47834763, 11.72975995, 12.12295317,  3.72734086, 12.55084872])

In [34]:
MSE=mean_squared_error(y,model.predict(X))
MSE

2.784126314510936

In [36]:
import numpy as np
RMSE=np.sqrt(MSE)
RMSE

1.6685701407225697

MODEL TUNING(MODEL DOĞRULAMA)

In [37]:
X.head()

TV  radio  newspaper
0  230.1   37.8       69.2
1   44.5   39.3       45.1
2   17.2   45.9       69.3
3  151.5   41.3       58.5
4  180.8   10.8       58.4

In [38]:
y.head()

0    22.1
1    10.4
2     9.3
3    18.5
4    12.9
Name: sales, dtype: float64

SINAMA SETİYLE 

In [39]:
#sınama seti
from sklearn.model_selection import train_test_split

In [40]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.20,random_state=99)#her çalıştırdığında farklı sonuçlar gelmesin diye random state kullanırız.

In [41]:
X_train.head()

TV  radio  newspaper
16    67.8   36.6      114.0
51   100.4    9.6        3.6
97   184.9   21.0       22.0
164  117.2   14.7        5.4
71   109.8   14.3       31.7

In [43]:
y_train.head()

16     12.5
51     10.7
97     15.5
164    11.9
71     12.4
Name: sales, dtype: float64

In [42]:
X_test.head()

TV  radio  newspaper
135  48.3   47.0        8.5
127  80.2    0.0        9.2
191  75.5   10.8        6.0
66   31.5   24.6        2.2
119  19.4   16.0       22.3

In [45]:
lm=LinearRegression()
model=lm.fit(X_train,y_train)

In [48]:
#eğitim hatası
np.sqrt(mean_squared_error(y_train,model.predict(X_train)))

1.7236824822650754

In [47]:
#test hatası
np.sqrt(mean_squared_error(y_test,model.predict(X_test)))

1.4312783138301641

Burada random state değiştikçe hata değişecek o yüzden bu yöntem yerine cross validation kullanmak daha mantıklı.

K KATLI CROSS VALIDATION

In [49]:
from sklearn.model_selection import cross_val_score

In [51]:
cross_val_score(model,X_train,y_train,cv=10,scoring="neg_mean_squared_error")
#scoring-->hata türü
##cv--> kaç katlı olamsını istiyosan.

array([-2.1019073 , -2.48953197, -3.09704214, -2.34694216, -3.68175761,
       -1.8691401 , -3.18173007, -4.1927349 , -2.17128376, -8.03821974])

In [53]:
np.mean(-cross_val_score(model,X_train,y_train,cv=10,scoring="neg_mean_squared_error"))
#ortalama hata
#cv rase için karekökü alınır.

3.3170289742341246